### PR TITLE
dtype support

### LINKED ISSUE #227: Numerical precision for high-resolution detectors
The test for numerical precision added in #226 uncovers that float32 may not be enough for 4096x4096 detectors to give sufficient numerical precision with standard reduce operations. We might have to use better algorithms like [Kahan summation](https://en.wikipedia.org/wiki/Kahan_summation_algorithm) or calculate with float64. 

For now, the test is "cheated" by using 1.0 as value, which doesn't give numerical errors.

### PR DESCRIPTION
**Pull request by uellue**
Tools and automatisms to control the dtype in which data is read and calculations are performed. All dtype combinations that support numpy.dot() should work now, including integer and complex number types for masks and/or data.

Closes #227 by allow switching to float64 or complex128.

Keep sparse.pydata.org for most of the sparse mask handling, but switch back to scipy.sparse for processing with the correct data layout, which is faster.

Drive-by change: Additional mask generation code and moving utility functions around.

### CODE REVIEW

File: tests/udf/test_blobfinder.py [hunk 1]
```diff
@@ -30,10 +31,12 @@ def test_refinenemt():
     assert (rx < x) and (rx > (x - 1))
 
 
+@pytest.mark.xfail(reason="FIXME update blobfinder UDF to new UDF interface")
```
Comment (sk1p) [inline tests/udf/test_blobfinder.py]: This needs to be removed; it is the cause for the decrease in coverage. I think you can just revert the whole file to the state it is in `master` and things should be fine.

**Comment by sk1p** [comment]
Other than the stray changes in the blobfinder test, this looks good!

File: tests/udf/test_blobfinder.py (hunk 1 above)
Comment (uellue) [reply tests/udf/test_blobfinder.py]: Done! :-)

**Comment by sk1p** [comment]
Could you rebase to current master? I'm now pinning `dask.distributed` to <1.28 to see if this helps with our CI hangs. Locally I couldn't reproduce the hangs with `1.27.1`, where it hung up quite consistently on `1.28`.

**Comment by uellue** [comment]
> Could you rebase to current master? I'm now pinning `dask.distributed` to <1.28 to see if this helps with our CI hangs. Locally I couldn't reproduce the hangs with `1.27.1`, where it hung up quite consistently on `1.28`.

Done!

**Comment by sk1p** [comment]
Looks good, thanks!

**Pull request merged**